pyccel/parser/semantic.py

```diff
@@ -2719,8 +2719,11 @@ def _handle_is_operator(self, IsClass, expr, **settings):
             (var2.is_Boolean or isinstance(var2.dtype, NativeBool))):
             return IsClass(var1, var2)
 
-        if ((var1.is_integer or isinstance(var1.dtype, NativeInteger)) and
-            (var2.is_integer or isinstance(var2.dtype, NativeInteger))):
+        lst = ["string", "complex", "real", "int"]
+        if (str(var1.dtype) in lst and str(var2.dtype) in lst):
+            errors.report(PYCCEL_RESTRICTION_PRIMITIVE_IMMUTABLE, symbol=expr,
+            bounding_box=(self._current_fst_node.lineno, self._current_fst_node.col_offset),
+            severity='error', blocker=self.blocking)
             return IsClass(var1, var2)
```

Comment:
I am OK with raising this error for an expression like `a is b`, but I think we can definitely handle the cases
```python
x is x      # True
x is not x  # False
```
where `x` is a **variable** of primitive immutable type, not an expression or a literal constant.

Reply:
I agree

Reply:
i agree